rust/src/domains/polars/expr/ffi.rs

```diff
@@ -0,0 +1,89 @@
+use std::ffi::c_char;
+
+use opendp_derive::bootstrap;
+
+use crate::{
+    core::{FfiResult, Metric, MetricSpace},
+    domains::LazyFrameDomain,
+    error::Fallible,
+    ffi::{
+        any::{AnyDomain, AnyMetric, AnyObject, Downcast},
+        util::{as_ref, to_option_str},
+    },
+    metrics::{InsertDeleteDistance, SymmetricDistance},
+};
+
+use super::{ExprContext, ExprDomain};
+
+#[no_mangle]
+#[bootstrap(
+    name = "expr_domain",
+    features("contrib"),
+    arguments(
+        lazyframe_domain(c_type = "AnyDomain *", rust_type = b"null"),
+        context(default = b"null", rust_type = b"null"),
+        grouping_columns(rust_type = "Option<Vec<String>>", default = b"null"),
+        active_column(
+            c_type = "AnyObject *",
+            rust_type = "Option<String>",
+            default = b"null"
+        )
+    )
+)]
+/// Construct an ExprDomain from a LazyFrameDomain.
+///
+/// Must pass either `context` or `grouping_columns`.
+///
+/// # Arguments
+/// * `lazyframe_domain` - the domain of the LazyFrame to be constructed
+/// * `context` - used when the constructor is called inside a lazyframe context constructor
+/// * `grouping_columns` - used when the constructor is called inside a groupby context constructor
+pub extern "C" fn opendp_domains__expr_domain(
+    lazyframe_domain: *const AnyDomain,
+    context: *const c_char,
+    grouping_columns: *const AnyObject,
+) -> FfiResult<*mut AnyDomain> {
+    let lf_domain = try_!(try_as_ref!(lazyframe_domain).downcast_ref::<LazyFrameDomain>()).clone();
+
+    let context = if let Some(context_str) = try_!(to_option_str(context)) {
+        match context_str.to_lowercase().as_str() {
+            "select" => ExprContext::Select,
+            "filter" => ExprContext::Filter,
+            "with_columns" => ExprContext::WithColumns,
+            _ => {
+                return err!(
+                FFI,
+                "context must be one of `select`, `filter` or `with_columns`, or pass `grouping_columns`"
```

Comment:
Should there be a check earlier that either `context` or `grouping_columns` but not both have been passed? This message could then probably be simplified:

```suggestion
                "context must be one of `select`, `filter` or `with_columns`"
```

Reply:
Or would it be simpler to just have two separate functions, in Rust and in Python? We could get rid of all the run-time error checking, and we wouldn't have to worry about Python users having to deal with Rust errors.

Reply:
Places where expressions can be used in Polars:

1. To build a sorting key
2. Fill null/fill NaN
3. Filter
4. Select
5. Group by keys
6. Group by agg
7. Join keys
8. With column
9. Explode
10. Drop nulls
11. Top/bottom k
12. Shift

For our use, I think these fall into four categories:
1. Not useful on their own for DP (11, 12)
2. Must be leaf nodes, like only col or lit (2, 5, 7, 9, 10)
3. Must be row-by-row (1, 3, 8)
4. Aggregates (4, 6)

We don't need Expression contexts for categories one or two, leaving only row-by-row and aggregates. I'll change the API to take an optional list of grouping keys. If set, then you are in category four (aggregates), otherwise in category three (row-by-row). When the grouping keys are empty, you have selection, otherwise groupby/agg.

Reply:
I forgot-- delineating how exactly the expression will be used is necessary for the member check.